text/1004-transactions-api.rst

```diff
@@ -0,0 +1,619 @@
+
+**Introduce** two methods on a connection pool and one on connection:
+
+.. code-block:: typescript
+
+    type TransactionBlock = (Transaction) => Promise<T>;
+
+    interface Pool {
+        retry<T>(block: TransactionBlock, options: RetryOptions): T;
+        raw_transaction<T>(action: TransactionBlock, options?: TransactionOptions): T;
+    }
+    class Connection {
+        raw_transaction<T>(action: TransactionBlock, options?: TransactionOptions): T;
+    }
+
+Introduce interface for making queries:
+
+.. code-block:: typescript
+
+    interface Executor {
+        async query(query: string, args: QueryArgs = null): Promise<Set>;
+        async queryOne(query: string, args: QueryArgs = null): Promise<any>;
+        async queryJSON(query: string, args: QueryArgs = null): Promise<string>;
+        async queryOneJSON(query: string, args: QueryArgs = null): Promise<string>;
+        async execute(query: string): Promise<void>;
+    }
+
+And implement interface by respective classes:
+
+.. code-block:: typescript
+
+   class Transaction implmements Executor {/*...*/}
+   class Connection implmements Executor {/*...*/}
+   interface Pool extends Executor {/*...*/}
+
+While removing inherent methods with the same name.
+
+Example of the recommended transaction API:
+
+.. code-block:: typescript
+
+    await pool.retry(tx => {
+        let val = await tx.fetch("...")
+        await tx.execute("...", process_value(val))
+    })
+
+Example using ``try_transaction``:
+
+    await pool.try_transaction(tx => {
+        let val = await tx.fetch("...")
+        await tx.execute("...", process_value(val))
+    })
+
+Note the new API is very similar to older ``transaction`` except the queries are
+executed using transaction object, instead of connection itself.
+
+**Deprecate** ``transaction()`` method::
+
+    `connection.transaction(f)` is deprecated. Use `pool.retry(f)`
+    (preferred) or `connection.try_transaction(f)`.
+
+The ``RetryOptions`` signature:
+
+.. code-block:: typescript
+
+    type DelayFn = (attempt: number) => number;
+    enum AttempsOption {
+        All,
+        NetworkError,
+        ConcurrentUpdate,
+        Deadlock,
+    }
+    class RetryOptions {
+        constructor({
+            attempts: number = 3,
+            transactionOptions?: TransactionOptions,
+        });
+        transactionOptions(TransactionOptions): RetryOptions;
+        attempts(
+            which: AttemptsOption,
+            attempts: number,
+            delay_ms?: DelayFn,
+        ): RetryOptions;
+    }
+
+The ``PoolOptions`` object receives ``retryOptions`` field:
+
+.. code-block:: typescript
+
+    interface PoolOptions {
+      // ...
+      retryOptions?: RetryOptions;
+    }
+
+Exceptions API
+``````````````
+
+Error hierarchy is amended by introducing ``TransientError``, ``NetworkError``
+and ``EarlyNetworkError`` with the following relationships:
+
+.. code-block:: typescript
+
+    class TransientError extends TransactionError {}
```

Comment:
I'd reference the https://github.com/edgedb/edgedb/blob/master/edb/api/errors.txt file.  And it looks like some of these exceptions are server-side and some are client side. Please sort them and inherit from appropriate base classes in `errors.txt`.

I'm also not sure why we need `TransientError`. Why is `ExecutionError` (already defined) not sufficient?

Reply:
```
class InvalidValueError(ExecutionError):
class IntegrityError(ExecutionError):
class TransactionError(ExecutionError):
```
First two must not be retried as far as I understand. There are also a couple of `TransactionError` errors that we must not retry:
```                                                                     
            raise errors.TransactionError('already in explicit transaction')                                                                
        raise errors.TransactionError(f'there is no {name!r} savepoint')                                                                 
            raise errors.TransactionError(f'there is no {name!r} savepoint')                                                                               
            raise errors.TransactionError('already in transaction')                                                                                        
            raise errors.TransactionError('cannot commit: not in transaction') 
```

Reply:
NetworkError is now inherited from ClientError.